File: js/test/parse_spec.js
import parse from '../src/parse';
import expect from 'expect.js';

describe('parser', () => {
  it('parses a number', () => {
    expect(parse('0').toString()).to.equal('(ide 0)');
  });

  it('parses an atom', () => {
    expect(parse('a').toString()).to.equal('(ide a)');
  });

  it('parses a lambda', () => {
    expect(parse('fun(x) 0').toString()).
      to.equal('(lambda (ide x) (ide 0))');
  });

  it('parses a cond', () => {
    expect(parse('if a then b else c').toString()).
      to.equal('(cond (ide a) (ide b) (ide c))');
  });

  it('parses a function application', () => {
    expect(parse('f(x)').toString()).
      to.equal('(apply (ide f) (ide x))');
  });

  it('parses a double function application', () => {
    expect(parse('max(x)(y)').toString()).
      to.equal('(apply (apply (ide max) (ide x)) (ide y))');
  });

  it ('parses a double function application inside a cond', () => {
    // This resolves an ambiguity: does the expression mean
    // (1)  (if a then b else max(x))(y) or
    // (2)  if a then b else (max(x)(y)) ?
    // Answer: (2)

    let expr  = parse('if a then b else max(x)(y)').toString();
    let right = parse('if a then b else (max(x)(y))').toString();
    let wrong = parse('(if a then b else max(x))(y)').toString();

    expect(expr).to.equal(right);
    expect(expr).not.to.equal(wrong);

    expect(expr).to.equal('(cond (ide a) (ide b) (apply (apply (ide max) (ide x)) (ide y)))');
  });

  it('parses a let expression', () => {
    expect(parse('let n = 3 in n').toString()).
      to.equal('(block (def n (ide 3)) (ide n))');
  });

  it ('parses a recursive let expression', () => {
    expect(parse('let rec f = fun(x) f(x) in f(0)').toString()).
      to.equal('(block (rec (def f (lambda (ide x) (apply (ide f) (ide x))))) (apply (ide f) (ide 0)))');
  });


});

File: hs/test/ParseSpec.hs
module ParseSpec where

import Test.Hspec
import Test.QuickCheck

import Parse
import ShowExpr
import ASTGen
import AST

import Debug.Trace

spec :: Spec
spec = do
  describe "parseFun" $ do
    it "parses an identifier" $ do
      parseFun "x" `shouldBe` Var "x"

    it "parses a number" $ do
      parseFun "0" `shouldBe` Num 0

    it "parses a cond" $ do
      parseFun "if happy then 42 else 666" `shouldBe`
        Cond (Var "happy") (Num 42) (Num 666)

    it "parses a lambda" $ do
      parseFun "fun(x) x" `shouldBe`
        Lambda "x" (Var "x")

    it "parses a function application" $ do
      parseFun "f(1)" `shouldBe` FunCall (Var "f") (Num 1)

    it "parses a double function application" $ do
      parseFun "f(1)(2)" `shouldBe`
        FunCall (FunCall (Var "f") (Num 1)) (Num 2)

    it "parses a comma-ized function application" $ do
      parseFun "f(1, 2)" `shouldBe` parseFun "f(1)(2)"

    it "parses a let..in" $ do
      parseFun "let a = 2 in succ(a)" `shouldBe`
        Block (Assign "a" (Num 2)) (FunCall (Var "succ") (Var "a"))

    it "parses a let rec..in" $ do
      parseFun "let rec f = f(0) in f" `shouldBe`
        Block (Rec (Assign "f" (FunCall (Var "f") (Num 0)))) (Var "f")

    it "parses a seq" $ do
      parseFun "let a=3 then b=4 in a" `shouldBe`
        Block (Seq (Assign "a" (Num 3)) (Assign "b" (Num 4))) (Var "a")

    it "parses a parenthesized decl" $ do
      parseFun "let (a=3 then b=4) in a" `shouldBe`
        parseFun "let a=3 then b=4 in a"

    it "parses bracketed expressions" $ do
      parseFun "(if sad then f else g)(0)" `shouldBe`
        FunCall (Cond (Var "sad") (Var "f") (Var "g")) (Num 0)

    it "parses arithmetic and boolean operations" $ do
      parseFun "a+1" `shouldBe`
        FunCall (FunCall (Var "plus") (Var "a")) (Num 1)
      parseFun "a-1" `shouldBe`
        FunCall (FunCall (Var "minus") (Var "a")) (Num 1)
      parseFun "a*2" `shouldBe`
        FunCall (FunCall (Var "times") (Var "a")) (Num 2)
      parseFun "a/2" `shouldBe`
        FunCall (FunCall (Var "div") (Var "a")) (Num 2)
      parseFun "-a" `shouldBe` FunCall (Var "negate") (Var "a")
      parseFun "!b" `shouldBe` FunCall (Var "not") (Var "b")
      parseFun "happy & youknowit" `shouldBe`
        FunCall (FunCall (Var "and") (Var "happy")) (Var "youknowit")
      parseFun "b | !b" `shouldBe`
        FunCall (FunCall (Var "or") (Var "b")) (FunCall (Var "not") (Var "b"))

      parseFun "a+b*c" `shouldBe`
        FunCall (FunCall (Var "plus") (Var "a"))
                (FunCall (FunCall (Var "times") (Var "b")) (Var "c"))

      let ab = (FunCall (FunCall (Var "times") (Var "a")) (Var "b"))
      parseFun "a*b+c" `shouldBe`
        FunCall (FunCall (Var "plus") ab) (Var "c")

    it "is the opposite of showExpr" $ property $
      -- \expr -> length (showExpr expr) < 100
      \expr -> parseFun (showExpr expr) == expr
